### DISCUSSION TITLE
Understanding how the autobuilder stuff works. Help or a whitepaper needed.

### DISCUSSION THREAD

**Discussion by WarrenReynolds** [mjansen4857/pathplanner#192] [Q&A] [answered]
Hi, 
We used pathplanner last year (2022) for following Holonomic paths on a swerve drive robot and absolutely loved it.  I'd like to personally thank Michael and everyone else that has contributed to this amazing project. 

I'm so excited about implementing the new features of PP for the 2023 season, I'm trying to think how I would build some sample c++ code that could utilise all the new features, so I've been digging through the source code trying to understand whats going on under the bonnet.

While I have been active in identifying bugs and making suggestions through out the development of the new features of path planner over the last 6 months I have to now say that I'm lost when I'm trying to understand what is happening at the nuts and bolts level.  There was a time when I understood what was going on, but with all the new changes to the WPILib commands and C++ syntax in general, I'm now very confused when I'm trying to follow what code would execute and when.

For example what is WrappedEventCommand?  Is this there to handle the ownership issue of what/who owns the commands in the eventMap, and to make sure that commands in the event map can be re-used with the whole std::move changing ownership/scope thing?

There is now so much code in the path planner source that is obviously very understandable to software engineers but very confusing to the what I'm calling the "average FRC coder"

Is there anyone out there that is willing to write a brief whitepaper on the nuts and bolts of how all the advanced features of path planner work?

Something real simple like.
In your robot init you create instances of 
a...
b....
c....

Then in your robot container you create instances of your commands and an instance of your command map and add the commands to the map.  This get passed to the autobuilder .....
eg Maybe some sample code that create an command map that has a couple of print commands in it that say print "This command was launched on the path", "This command was launched at a stop point" or something along those lines. 

Then in your autochooser code you do this .......

I'm just worried that if I'm struggling to keep up when I have been following along reasonably closely, I don't think the average FRC student coding group is going to be able to utilise the new advanced features on this amazing piece of open source software.

**Comment by mjansen4857** [comment, marked as answer]
> For example what is WrappedEventCommand? Is this there to handle the ownership issue of what/who owns the commands in the eventMap, and to make sure that commands in the event map can be re-used with the whole std::move changing ownership/scope thing?

Just to answer this question, WrappedEventCommand is to get around some of the restrictions of command groups. For example, commands can only be added to a command group once. But there is only 1 instance of each command in the event map. So WrappedEventCommand is used to create a "unique" command that just calls the methods in the original command under the hood.

Anyways, it sounds like you're asking for an example project over some description of how everything works internally? I do think a full AutoBuilder example project would be quite useful so I can get on that. As for how it works internally, I can't really go into too much depth, because it is fairly simple as long as you are pretty familiar with the WPILib commands framework. You'll likely benefit more from looking more into that. 

But, a basic outline of what it does:
1. Create a sequential command group as the root command of the full auto
2. Add a command to reset the robot pose for the beginning of the auto
3. For every path in the path group:
    1. Add a command group for the stop event at the beginning of this path to the root command group. Note: Start stop event of path 2 = end stop event of path 1
        * The type of this command group is determined by the "execution behavior". Every command that should be part of this group is added as a WrappedEventCommand
            * Sequential = SequentialCommandGroup
            * Parallel = ParallelCommandGroup
            * Deadline = ParallelDeadlineGroup. 
        * The above group is then added to another command group based on the wait behavior. 
            * Before = Sequential group (wait then events)
            * After = Sequential group (events then wait)
            * Deadline = Deadline group with the wait as the deadline
            * Minimum = Parallel group
            * None = no wait, just use the above event group
    2. Add a FollowPathWithEvents command for this path to the root command group
        * Uses different path following commands based on the type of auto builder
        * Works like a parallel deadline group with the path following command as a deadline
        * As the path following group runs, a timer is used to start running event marker commands when they are supposed to be triggered
4. Add a command group for the stop event at the end of the last path in the path group to the root command group. Same as the stop event group described above

**Reply by WarrenReynolds** [reply]
That explains it perfectly.

**Reply by WarrenReynolds** [reply]
Only real question I have now is the c++ syntax of adding commands to the event map.  Would you create an instance on the commands as a member variable to the robot container and then add these pointer to the map? Then their scope is the lifetime of the robotcontainer (which is forever)?

**Reply by mjansen4857** [reply]
The commands won’t be destroyed as long as the auto builder remains in scope. The auto builder holds the map so it will keep the commands in scope.

**Comment by WarrenReynolds** [comment]
Hi,
Great work on updating the wiki.  I noticed that you have added a lot of sample code snippets to the wiki and these helped our team immensely today. 

I think I may have found a typo in the C++ AutoBuilder section.

In the example code for creating an auto builder, I think the parameter to supply the subsystem should be changed from { swerveSubsystem } to { &swerveSubsystem } as the constructor for an AutoBuilder is expecting a
std::initializer_list<<frc2::Subsystem*>> 

I could be wrong, as I don't fully understand std:initializer_list, but we found that we had to change this today to get our code to compile.

Looking forward to testing the full auto builder in the coming days.

Cheers
Warren

**Reply by mjansen4857** [reply]
Yeah, you're right. I'll fix that.
